Attachment Refactor for Client

AlexF301 opened the pull request:
- Refactored perform_attachment_job() in Client.py to not use a try-except block. Refactor checks if the JSON for an attachment is valid, essentially checking if there are any attachments to download. If there aren't attachments, no further steps need to be taken. 

- Also added some mock tests that mimic the handling of TypeError (NoneType) and IndexError

KyleSmilon commented:
This looks good, am I correct in seeing that if it is None-type or index error it will print `FAILURE: ...` for whichever case within the `is_attachment_json_valid`. @AlexF301

jack11wagner commented:
I agree with Kyle here, maybe just printing out no attachments to download would be better.

AlexF301 commented:
Updated part of the code to reflect feedback provided

- Changed print statement to print out that no attachments are available to download
- changed function name to improve readability
- removed parameter which had no benefit to function

Just realized I left some print statements on there I was using to debug. Those will be removed in next push

Not ready to merge

nikovacs commented on mirrulations-client/src/mirrclient/client.py:
```diff
@@ -236,30 +237,41 @@ def perform_attachment_job(self, url, job_id):
+            return False
+        # Handles NoneType
+        if len(attachment_json['data']) >= 1:
+            if not attachment_json['data'][0]['attributes']['fileFormats']:
```
It would be best if we also checked for the existence of `attributes` and `fileFormats` before attempting to retrieve them from the JSON.

nikovacs commented on mirrulations-client/src/mirrclient/client.py:
```diff
@@ -236,30 +237,41 @@ def perform_attachment_job(self, url, job_id):
+        True or False depending if there is an attachment available to download
+        """
+        # handle keyError
+        if "data" not in attachment_json:
```
This if statement and the second if statement could be combined. For example, the second expression could be something like `attachment_json.get('data') in (None, [])`

nikovacs replied:
Or `attachment_json.get('data', []) == []`

nikovacs reviewed (commented):
Looks good, but could use the two small changes mentioned in my review.

bjcoleman commented:
@nikovacs We aren't going to get 'fileFormats` at this point - just the attachments on comments.  We'll do another round with the data to do other places where attachments are reference.

nikovacs commented:
> @nikovacs We aren't going to get 'fileFormats` at this point - just the attachments on comments. We'll do another round with the data to do other places where attachments are reference.

The retrieval of 'fileFormats' in this method already existed prior to our work on this project. This PR is eliminating the use of try/except while also being less prone to error.

nikovacs commented:
Looks good, merging.
